<a href="https://colab.research.google.com/github/xsaravix/TelecomX/blob/main/TelecomX.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#📌 Extracão

Importando as bibliotecas


In [ ]:
import pandas as pd
import requests
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.graph_objects as go

Leitura, normalização e limpeza dos dados

In [ ]:
url = 'https://raw.githubusercontent.com/alura-cursos/challenge2-data-science/refs/heads/main/TelecomX_Data.json'

In [ ]:
response = requests.get(url)
data = response.json()

dados = pd.DataFrame(data)

In [ ]:
dados.head()

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

In [ ]:
from pandas import json_normalize

dados_normalizados = json_normalize(data)
dados_normalizados.head()

customerID Churn customer.gender  customer.SeniorCitizen customer.Partner  \
0  0002-ORFBO    No          Female                       0              Yes   
1  0003-MKNFE    No            Male                       0               No   
2  0004-TLHLJ   Yes            Male                       0               No   
3  0011-IGKFF   Yes            Male                       1              Yes   
4  0013-EXCHZ   Yes          Female                       1              Yes   

  customer.Dependents  customer.tenure phone.PhoneService phone.MultipleLines  \
0                 Yes                9                Yes                  No   
1                  No                9                Yes                 Yes   
2                  No                4                Yes                  No   
3                  No               13                Yes                  No   
4                  No                3                Yes                  No   

  internet.InternetService  ... internet.OnlineBackup  \
0                      DSL  ...                   Yes   
1                      DSL  ...                    No   
2              Fiber optic  ...                    No   
3              Fiber optic  ...                   Yes   
4              Fiber optic  ...                    No   

  internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
0                        No                  Yes                  Yes   
1                        No                   No                   No   
2                       Yes                   No                   No   
3                       Yes                   No                  Yes   
4                        No                  Yes                  Yes   

  internet.StreamingMovies account.Contract account.PaperlessBilling  \
0                       No         One year                      Yes   
1                      Yes   Month-to-month                       No   
2                       No   Month-to-month                      Yes   
3                      Yes   Month-to-month                      Yes   
4                       No   Month-to-month                      Yes   

  account.PaymentMethod account.Charges.Monthly  account.Charges.Total  
0          Mailed check                    65.6                  593.3  
1          Mailed check                    59.9                  542.4  
2      Electronic check                    73.9                 280.85  
3      Electronic check                    98.0                1237.85  
4          Mailed check                    83.9                  267.4  

[5 rows x 21 columns]

#🔧 Transformação

Verificando inconsistências

In [ ]:
dados_normalizados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer.gender            7267 non-null   object 
 3   customer.SeniorCitizen     7267 non-null   int64  
 4   customer.Partner           7267 non-null   object 
 5   customer.Dependents        7267 non-null   object 
 6   customer.tenure            7267 non-null   int64  
 7   phone.PhoneService         7267 non-null   object 
 8   phone.MultipleLines        7267 non-null   object 
 9   internet.InternetService   7267 non-null   object 
 10  internet.OnlineSecurity    7267 non-null   object 
 11  internet.OnlineBackup      7267 non-null   object 
 12  internet.DeviceProtection  7267 non-null   object 
 13  internet.TechSupport       7267 non-null   objec

In [ ]:
dados_normalizados.isnull().sum()

customerID                   0
Churn                        0
customer.gender              0
customer.SeniorCitizen       0
customer.Partner             0
customer.Dependents          0
customer.tenure              0
phone.PhoneService           0
phone.MultipleLines          0
internet.InternetService     0
internet.OnlineSecurity      0
internet.OnlineBackup        0
internet.DeviceProtection    0
internet.TechSupport         0
internet.StreamingTV         0
internet.StreamingMovies     0
account.Contract             0
account.PaperlessBilling     0
account.PaymentMethod        0
account.Charges.Monthly      0
account.Charges.Total        0
dtype: int64

In [ ]:
for coluna in dados_normalizados.columns:
  print(f'{coluna}: {dados_normalizados[coluna].unique()}')

customerID: ['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH']
Churn: ['No' 'Yes' '']
customer.gender: ['Female' 'Male']
customer.SeniorCitizen: [0 1]
customer.Partner: ['Yes' 'No']
customer.Dependents: ['Yes' 'No']
customer.tenure: [ 9  4 13  3 71 63  7 65 54 72  5 56 34  1 45 50 23 55 26 69 11 37 49 66
 67 20 43 59 12 27  2 25 29 14 35 64 39 40  6 30 70 57 58 16 32 33 10 21
 61 15 44 22 24 19 47 62 46 52  8 60 48 28 41 53 68 51 31 36 17 18 38 42
  0]
phone.PhoneService: ['Yes' 'No']
phone.MultipleLines: ['No' 'Yes' 'No phone service']
internet.InternetService: ['DSL' 'Fiber optic' 'No']
internet.OnlineSecurity: ['No' 'Yes' 'No internet service']
internet.OnlineBackup: ['Yes' 'No' 'No internet service']
internet.DeviceProtection: ['No' 'Yes' 'No internet service']
internet.TechSupport: ['Yes' 'No' 'No internet service']
internet.StreamingTV: ['Yes' 'No' 'No internet service']
internet.StreamingMovies: ['No' 'Yes' 'No internet service']
account.Contrac

In [ ]:
dados_normalizados.duplicated().sum()

np.int64(0)

In [ ]:
dados_normalizados.apply(lambda x: x.astype(str).str.strip() == '').sum()

customerID                     0
Churn                        224
customer.gender                0
customer.SeniorCitizen         0
customer.Partner               0
customer.Dependents            0
customer.tenure                0
phone.PhoneService             0
phone.MultipleLines            0
internet.InternetService       0
internet.OnlineSecurity        0
internet.OnlineBackup          0
internet.DeviceProtection      0
internet.TechSupport           0
internet.StreamingTV           0
internet.StreamingMovies       0
account.Contract               0
account.PaperlessBilling       0
account.PaymentMethod          0
account.Charges.Monthly        0
account.Charges.Total         11
dtype: int64

Tratando as inconsistências

In [ ]:
dados_normalizados.columns = dados_normalizados.columns.str.replace('.', '_', regex=False)

In [ ]:
dados_normalizados.head()

customerID Churn customer_gender  customer_SeniorCitizen customer_Partner  \
0  0002-ORFBO    No          Female                       0              Yes   
1  0003-MKNFE    No            Male                       0               No   
2  0004-TLHLJ   Yes            Male                       0               No   
3  0011-IGKFF   Yes            Male                       1              Yes   
4  0013-EXCHZ   Yes          Female                       1              Yes   

  customer_Dependents  customer_tenure phone_PhoneService phone_MultipleLines  \
0                 Yes                9                Yes                  No   
1                  No                9                Yes                 Yes   
2                  No                4                Yes                  No   
3                  No               13                Yes                  No   
4                  No                3                Yes                  No   

  internet_InternetService  ... internet_OnlineBackup  \
0                      DSL  ...                   Yes   
1                      DSL  ...                    No   
2              Fiber optic  ...                    No   
3              Fiber optic  ...                   Yes   
4              Fiber optic  ...                    No   

  internet_DeviceProtection internet_TechSupport internet_StreamingTV  \
0                        No                  Yes                  Yes   
1                        No                   No                   No   
2                       Yes                   No                   No   
3                       Yes                   No                  Yes   
4                        No                  Yes                  Yes   

  internet_StreamingMovies account_Contract account_PaperlessBilling  \
0                       No         One year                      Yes   
1                      Yes   Month-to-month                       No   
2                       No   Month-to-month                      Yes   
3                      Yes   Month-to-month                      Yes   
4                       No   Month-to-month                      Yes   

  account_PaymentMethod account_Charges_Monthly  account_Charges_Total  
0          Mailed check                    65.6                  593.3  
1          Mailed check                    59.9                  542.4  
2      Electronic check                    73.9                 280.85  
3      Electronic check                    98.0                1237.85  
4          Mailed check                    83.9                  267.4  

[5 rows x 21 columns]

In [ ]:
churn_desconhecido = dados_normalizados.query('Churn == ""')

In [ ]:
churn_desconhecido.head()

customerID Churn customer_gender  customer_SeniorCitizen  \
30   0047-ZHDTW                Female                       0   
75   0120-YZLQA                  Male                       0   
96   0154-QYHJU                  Male                       0   
98   0162-RZGMZ                Female                       1   
175  0274-VVQOQ                  Male                       1   

    customer_Partner customer_Dependents  customer_tenure phone_PhoneService  \
30                No                  No               11                Yes   
75                No                  No               71                Yes   
96                No                  No               29                Yes   
98                No                  No                5                Yes   
175              Yes                  No               65                Yes   

    phone_MultipleLines internet_InternetService  ... internet_OnlineBackup  \
30                  Yes              Fiber optic  ...                    No   
75                   No                       No  ...   No internet service   
96                   No                      DSL  ...                   Yes   
98                   No                      DSL  ...                   Yes   
175                 Yes              Fiber optic  ...                   Yes   

    internet_DeviceProtection internet_TechSupport internet_StreamingTV  \
30                         No                   No                   No   
75        No internet service  No internet service  No internet service   
96                         No                  Yes                   No   
98                         No                  Yes                   No   
175                       Yes                   No                  Yes   

    internet_StreamingMovies account_Contract account_PaperlessBilling  \
30                        No   Month-to-month                      Yes   
75       No internet service         Two year                      Yes   
96                        No         One year                      Yes   
98                        No   Month-to-month                       No   
175                      Yes         One year                      Yes   

         account_PaymentMethod account_Charges_Monthly  account_Charges_Total  
30   Bank transfer (automatic)                   79.00                  929.3  
75     Credit card (automatic)                   19.90                 1355.1  
96            Electronic check                   58.75                 1696.2  
98     Credit card (automatic)                   59.90                 287.85  
175  Bank transfer (automatic)                  103.15                6792.45  

[5 rows x 21 columns]

In [ ]:
churn_desconhecido.tail()

customerID Churn customer_gender  customer_SeniorCitizen  \
7158  9840-GSRFX                Female                       0   
7180  9872-RZQQB                Female                       0   
7211  9920-GNDMB                  Male                       0   
7239  9955-RVWSC                Female                       0   
7247  9966-VYRTZ                Female                       0   

     customer_Partner customer_Dependents  customer_tenure phone_PhoneService  \
7158               No                  No               14                Yes   
7180              Yes                  No               49                 No   
7211               No                  No                9                Yes   
7239              Yes                 Yes               67                Yes   
7247              Yes                 Yes               31                Yes   

     phone_MultipleLines internet_InternetService  ... internet_OnlineBackup  \
7158                 Yes                      DSL  ...                   Yes   
7180    No phone service                      DSL  ...                    No   
7211                 Yes              Fiber optic  ...                    No   
7239                  No                       No  ...   No internet service   
7247                  No                       No  ...   No internet service   

     internet_DeviceProtection internet_TechSupport internet_StreamingTV  \
7158                        No                   No                   No   
7180                        No                   No                  Yes   
7211                        No                   No                   No   
7239       No internet service  No internet service  No internet service   
7247       No internet service  No internet service  No internet service   

     internet_StreamingMovies account_Contract account_PaperlessBilling  \
7158                       No         One year                      Yes   
7180                       No   Month-to-month                       No   
7211                       No   Month-to-month                      Yes   
7239      No internet service         Two year                      Yes   
7247      No internet service   Month-to-month                      Yes   

          account_PaymentMethod account_Charges_Monthly  account_Charges_Total  
7158               Mailed check                   54.25                  773.2  
7180  Bank transfer (automatic)                   40.65                2070.75  
7211           Electronic check                   76.25                 684.85  
7239  Bank transfer (automatic)                   19.25                 1372.9  
7247               Mailed check                   19.55                 658.95  

[5 rows x 21 columns]

In [ ]:
# Removendo as linhas onde Churn está sem valor
dados_normalizados = dados_normalizados[dados_normalizados['Churn'].str.strip() != '']

In [ ]:
#Alterando a coluna "account.Charges.Total" para Float

dados_normalizados['account_Charges_Total'] = pd.to_numeric(dados_normalizados['account_Charges_Total'], errors='coerce')
dados_normalizados['account_Charges_Total'].dtype

/tmp/ipython-input-17-2795916313.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dados_normalizados['account_Charges_Total'] = pd.to_numeric(dados_normalizados['account_Charges_Total'], errors='coerce')


dtype('float64')

In [ ]:
dados_normalizados.apply(lambda x: x.astype(str).str.strip() == '').sum()

customerID                   0
Churn                        0
customer_gender              0
customer_SeniorCitizen       0
customer_Partner             0
customer_Dependents          0
customer_tenure              0
phone_PhoneService           0
phone_MultipleLines          0
internet_InternetService     0
internet_OnlineSecurity      0
internet_OnlineBackup        0
internet_DeviceProtection    0
internet_TechSupport         0
internet_StreamingTV         0
internet_StreamingMovies     0
account_Contract             0
account_PaperlessBilling     0
account_PaymentMethod        0
account_Charges_Monthly      0
account_Charges_Total        0
dtype: int64

Analisando colunas para entender a importância das variáveis na análise e fazer possíveis remoções

In [ ]:
dados_normalizados

customerID Churn customer_gender  customer_SeniorCitizen  \
0     0002-ORFBO    No          Female                       0   
1     0003-MKNFE    No            Male                       0   
2     0004-TLHLJ   Yes            Male                       0   
3     0011-IGKFF   Yes            Male                       1   
4     0013-EXCHZ   Yes          Female                       1   
...          ...   ...             ...                     ...   
7262  9987-LUTYD    No          Female                       0   
7263  9992-RRAMN   Yes            Male                       0   
7264  9992-UJOEL    No            Male                       0   
7265  9993-LHIEB    No            Male                       0   
7266  9995-HOTOH    No            Male                       0   

     customer_Partner customer_Dependents  customer_tenure phone_PhoneService  \
0                 Yes                 Yes                9                Yes   
1                  No                  No                9                Yes   
2                  No                  No                4                Yes   
3                 Yes                  No               13                Yes   
4                 Yes                  No                3                Yes   
...               ...                 ...              ...                ...   
7262               No                  No               13                Yes   
7263              Yes                  No               22                Yes   
7264               No                  No                2                Yes   
7265              Yes                 Yes               67                Yes   
7266              Yes                 Yes               63                 No   

     phone_MultipleLines internet_InternetService  ... internet_OnlineBackup  \
0                     No                      DSL  ...                   Yes   
1                    Yes                      DSL  ...                    No   
2                     No              Fiber optic  ...                    No   
3                     No              Fiber optic  ...                   Yes   
4                     No              Fiber optic  ...                    No   
...                  ...                      ...  ...                   ...   
7262                  No                      DSL  ...                    No   
7263                 Yes              Fiber optic  ...                    No   
7264                  No                      DSL  ...                   Yes   
7265                  No                      DSL  ...                    No   
7266    No phone service                      DSL  ...                   Yes   

     internet_DeviceProtection internet_TechSupport internet_StreamingTV  \
0                           No                  Yes                  Yes   
1                           No                   No                   No   
2                          Yes                   No                   No   
3                          Yes                   No                  Yes   
4                           No                  Yes                  Yes   
...                        ...                  ...                  ...   
7262                        No                  Yes                   No   
7263                        No                   No                   No   
7264                        No                   No                   No   
7265                       Yes                  Yes                   No   
7266                       Yes                   No                  Yes   

     internet_StreamingMovies account_Contract account_PaperlessBilling  \
0                          No         One year                      Yes   
1                         Yes   Month-to-month                       No   
2                          No   Month-to-month                      Yes   
3                         Yes   Month-to-month                    

In [ ]:
dados_normalizados = dados_normalizados.drop(['customerID'], axis=1)

In [ ]:
dados_normalizados.reset_index(drop=True, inplace=True)

In [ ]:
dados_normalizados = dados_normalizados.replace('Yes', 1)
dados_normalizados = dados_normalizados.replace('No', 0)

/tmp/ipython-input-22-2734015794.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  dados_normalizados = dados_normalizados.replace('No', 0)


In [ ]:
dados_normalizados.head()

Churn customer_gender  customer_SeniorCitizen  customer_Partner  \
0      0          Female                       0                 1   
1      0            Male                       0                 0   
2      1            Male                       0                 0   
3      1            Male                       1                 1   
4      1          Female                       1                 1   

   customer_Dependents  customer_tenure  phone_PhoneService  \
0                    1                9                   1   
1                    0                9                   1   
2                    0                4                   1   
3                    0               13                   1   
4                    0                3                   1   

  phone_MultipleLines internet_InternetService internet_OnlineSecurity  \
0                   0                      DSL                       0   
1                   1                      DSL                       0   
2                   0              Fiber optic                       0   
3                   0              Fiber optic                       0   
4                   0              Fiber optic                       0   

  internet_OnlineBackup internet_DeviceProtection internet_TechSupport  \
0                     1                         0                    1   
1                     0                         0                    0   
2                     0                         1                    0   
3                     1                         1                    0   
4                     0                         0                    1   

  internet_StreamingTV internet_StreamingMovies account_Contract  \
0                    1                        0         One year   
1                    0                        1   Month-to-month   
2                    0                        0   Month-to-month   
3                    1                        1   Month-to-month   
4                    1                        0   Month-to-month   

   account_PaperlessBilling account_PaymentMethod  account_Charges_Monthly  \
0                         1          Mailed check                     65.6   
1                         0          Mailed check                     59.9   
2                         1      Electronic check                     73.9   
3                         1      Electronic check                     98.0   
4                         1          Mailed check                     83.9   

   account_Charges_Total  
0                 593.30  
1                 542.40  
2                 280.85  
3                1237.85  
4                 267.40

In [ ]:
dados_normalizados['contas_diarias'] = (dados_normalizados['account_Charges_Monthly'] / 30).round(2)
dados_normalizados.head()

Churn customer_gender  customer_SeniorCitizen  customer_Partner  \
0      0          Female                       0                 1   
1      0            Male                       0                 0   
2      1            Male                       0                 0   
3      1            Male                       1                 1   
4      1          Female                       1                 1   

   customer_Dependents  customer_tenure  phone_PhoneService  \
0                    1                9                   1   
1                    0                9                   1   
2                    0                4                   1   
3                    0               13                   1   
4                    0                3                   1   

  phone_MultipleLines internet_InternetService internet_OnlineSecurity  ...  \
0                   0                      DSL                       0  ...   
1                   1                      DSL                       0  ...   
2                   0              Fiber optic                       0  ...   
3                   0              Fiber optic                       0  ...   
4                   0              Fiber optic                       0  ...   

  internet_DeviceProtection internet_TechSupport internet_StreamingTV  \
0                         0                    1                    1   
1                         0                    0                    0   
2                         1                    0                    0   
3                         1                    0                    1   
4                         0                    1                    1   

  internet_StreamingMovies account_Contract account_PaperlessBilling  \
0                        0         One year                        1   
1                        1   Month-to-month                        0   
2                        0   Month-to-month                        1   
3                        1   Month-to-month                        1   
4                        0   Month-to-month                        1   

   account_PaymentMethod account_Charges_Monthly  account_Charges_Total  \
0           Mailed check                    65.6                 593.30   
1           Mailed check                    59.9                 542.40   
2       Electronic check                    73.9                 280.85   
3       Electronic check                    98.0                1237.85   
4           Mailed check                    83.9                 267.40   

   contas_diarias  
0            2.19  
1            2.00  
2            2.46  
3            3.27  
4            2.80  

[5 rows x 21 columns]

In [ ]:
dados_normalizados = dados_normalizados.rename(columns={'customer_gender': 'genero', 'customer_SeniorCitizen': 'idoso', 'customer_Partner': 'parceiro',
                                                        'customer_Dependents': 'dependentes', 'customer_tenure': 'meses_contrato', 'phone_PhoneService': 'servico_telefonico',
                                                        'phone_MultipleLines': 'multiplas_linhas', 'internet_InternetService': 'servico_internet',
                                                        'internet_OnlineSecurity': 'seguranca_online', 'internet_OnlineBackup': 'backup_online',
                                                        'internet_DeviceProtection': 'protecao_dispositivo', 'internet_TechSupport': 'suporte_tecnico',
                                                        'internet_StreamingTV': 'tv_a_cabo', 'internet_StreamingMovies': 'streaming_filmes',
                                                        'account_Contract': 'contrato', 'account_PaperlessBilling': 'fatura_online', 'account_PaymentMethod': 'metodo_pagamento',
                                                        'account_Charges_Monthly': 'cobranca_mensal', 'account_Charges_Total': 'gasto_total'})
dados_normalizados.head(10)

Churn  genero  idoso  parceiro  dependentes  meses_contrato  \
0      0  Female      0         1            1               9   
1      0    Male      0         0            0               9   
2      1    Male      0         0            0               4   
3      1    Male      1         1            0              13   
4      1  Female      1         1            0               3   
5      0  Female      0         0            1               9   
6      0  Female      1         1            0              71   
7      0    Male      0         1            0              63   
8      0  Female      1         0            0               7   
9      0  Female      0         1            1              65   

   servico_telefonico multiplas_linhas servico_internet seguranca_online  ...  \
0                   1                0              DSL                0  ...   
1                   1                1              DSL                0  ...   
2                   1                0      Fiber optic                0  ...   
3                   1                0      Fiber optic                0  ...   
4                   1                0      Fiber optic                0  ...   
5                   1                0              DSL                0  ...   
6                   1                0      Fiber optic                1  ...   
7                   1                1      Fiber optic                1  ...   
8                   1                0              DSL                1  ...   
9                   1                1              DSL                1  ...   

  protecao_dispositivo suporte_tecnico tv_a_cabo streaming_filmes  \
0                    0               1         1                0   
1                    0               0         0                1   
2                    1               0         0                0   
3                    1               0         1                1   
4                    0               1         1                0   
5                    0               1         1                1   
6                    1               1         1                1   
7                    0               1         0                0   
8                    0               0         0                0   
9                    1               1         1                1   

         contrato fatura_online           metodo_pagamento cobranca_mensal  \
0        One year             1               Mailed check           65.60   
1  Month-to-month             0               Mailed check           59.90   
2  Month-to-month             1           Electronic check           73.90   
3  Month-to-month             1           Electronic check           98.00   
4  Month-to-month             1               Mailed check           83.90   
5  Month-to-month             1    Credit card (automatic)           69.40   
6        Two year             1  Bank transfer (automatic)          109.70   
7        Two year             1    Credit card (automatic)           84.65   
8  Month-to-month             1           Electronic check           48.20   
9        Two year             1               Mailed check           90.45   

   gasto_total  contas_diarias  
0       593.30            2.19  
1       542.40            2.00  
2       280.85            2.46  
3      1237.85            3.27  
4       267.40            2.80  
5       571.45            2.31  
6      7904.25            3.66  
7      5377.80            2.82  
8       340.35            1.61  
9      5957.90            3.02  

[10 rows x 21 columns]

In [ ]:
substituicoes = {
    'Male': 'masculino',
    'Female': 'feminino',
    'Month-to-month': 'mensal',
    'One year': 'anual',
    'Two year': 'bienal',
    'Fiber optic': 'fibra óptica',
    'Mailed check': 'cheque',
    'Electronic check': 'cheque eletrônico',
    'Credit card (automatic)': 'cartão de crédito automático',
    'Bank transfer (automatic)': 'transferência bancária automática',
    'No phone service': 'sem serviço telefônico',
    'No internet service': 'sem serviço de internet'
}

dados_normalizados.replace(substituicoes, inplace=True)

In [ ]:
dados_normalizados.head()

Churn     genero  idoso  parceiro  dependentes  meses_contrato  \
0      0   feminino      0         1            1               9   
1      0  masculino      0         0            0               9   
2      1  masculino      0         0            0               4   
3      1  masculino      1         1            0              13   
4      1   feminino      1         1            0               3   

   servico_telefonico multiplas_linhas servico_internet seguranca_online  ...  \
0                   1                0              DSL                0  ...   
1                   1                1              DSL                0  ...   
2                   1                0     fibra óptica                0  ...   
3                   1                0     fibra óptica                0  ...   
4                   1                0     fibra óptica                0  ...   

  protecao_dispositivo suporte_tecnico tv_a_cabo streaming_filmes contrato  \
0                    0               1         1                0    anual   
1                    0               0         0                1   mensal   
2                    1               0         0                0   mensal   
3                    1               0         1                1   mensal   
4                    0               1         1                0   mensal   

  fatura_online   metodo_pagamento cobranca_mensal  gasto_total  \
0             1             cheque            65.6       593.30   
1             0             cheque            59.9       542.40   
2             1  cheque eletrônico            73.9       280.85   
3             1  cheque eletrônico            98.0      1237.85   
4             1             cheque            83.9       267.40   

   contas_diarias  
0            2.19  
1            2.00  
2            2.46  
3            3.27  
4            2.80  

[5 rows x 21 columns]

In [ ]:
dados_normalizados = dados_normalizados.applymap(lambda x: x.lower() if isinstance(x, str) else x)
dados_normalizados

/tmp/ipython-input-28-1951704782.py:1: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  dados_normalizados = dados_normalizados.applymap(lambda x: x.lower() if isinstance(x, str) else x)


Churn     genero  idoso  parceiro  dependentes  meses_contrato  \
0         0   feminino      0         1            1               9   
1         0  masculino      0         0            0               9   
2         1  masculino      0         0            0               4   
3         1  masculino      1         1            0              13   
4         1   feminino      1         1            0               3   
...     ...        ...    ...       ...          ...             ...   
7038      0   feminino      0         0            0              13   
7039      1  masculino      0         1            0              22   
7040      0  masculino      0         0            0               2   
7041      0  masculino      0         1            1              67   
7042      0  masculino      0         1            1              63   

      servico_telefonico        multiplas_linhas servico_internet  \
0                      1                       0              dsl   
1                      1                       1              dsl   
2                      1                       0     fibra óptica   
3                      1                       0     fibra óptica   
4                      1                       0     fibra óptica   
...                  ...                     ...              ...   
7038                   1                       0              dsl   
7039                   1                       1     fibra óptica   
7040                   1                       0              dsl   
7041                   1                       0              dsl   
7042                   0  sem serviço telefônico              dsl   

     seguranca_online  ... protecao_dispositivo suporte_tecnico tv_a_cabo  \
0                   0  ...                    0               1         1   
1                   0  ...                    0               0         0   
2                   0  ...                    1               0         0   
3                   0  ...                    1               0         1   
4                   0  ...                    0               1         1   
...               ...  ...                  ...             ...       ...   
7038                1  ...                    0               1         0   
7039                0  ...                    0               0         0   
7040                0  ...                    0               0         0   
7041                1  ...                    1               1         0   
7042                1  ...                    1               0         1   

     streaming_filmes contrato fatura_online   metodo_pagamento  \
0                   0    anual             1             cheque   
1                   1   mensal             0             cheque   
2                   0   mensal             1  cheque eletrônico   
3                   1   mensal             1  cheque eletrônico   
4                   0   mensal             1             cheque   
...               ...      ...           ...                ...   
7038                0    anual             0             cheque   
7039                1   mensal             1  cheque eletrônico   
7040                0   mensal             1             cheque   
7041                1   bienal             0             cheque   
7042                1   bienal             0  cheque eletrônico   

     cobranca_mensal  gasto_total  contas_diarias  
0              65.60       593.30            2.19  
1              59.90       542.40            2.00  
2              73.90       280.85            2.46  
3              98.00      1237.85            3.27  
4              83.90       267.40            2.80  
...              ...          ...             ...  
7038           55.15       742.90            1.84  
7039           85.10      1873.70            2.84  
7040           50.30        92.75            1.68  
7041           67.85      4627.65            2.26  
7042      

#📊 Carga e análise

In [ ]:
dados_normalizados.describe()

Churn        idoso     parceiro  dependentes  meses_contrato  \
count  7043.000000  7043.000000  7043.000000  7043.000000     7043.000000   
mean      0.265370     0.162147     0.483033     0.299588       32.371149   
std       0.441561     0.368612     0.499748     0.458110       24.559481   
min       0.000000     0.000000     0.000000     0.000000        0.000000   
25%       0.000000     0.000000     0.000000     0.000000        9.000000   
50%       0.000000     0.000000     0.000000     0.000000       29.000000   
75%       1.000000     0.000000     1.000000     1.000000       55.000000   
max       1.000000     1.000000     1.000000     1.000000       72.000000   

       servico_telefonico  fatura_online  cobranca_mensal  gasto_total  \
count         7043.000000    7043.000000      7043.000000  7032.000000   
mean             0.903166       0.592219        64.761692  2283.300441   
std              0.295752       0.491457        30.090047  2266.771362   
min              0.000000       0.000000        18.250000    18.800000   
25%              1.000000       0.000000        35.500000   401.450000   
50%              1.000000       1.000000        70.350000  1397.475000   
75%              1.000000       1.000000        89.850000  3794.737500   
max              1.000000       1.000000       118.750000  8684.800000   

       contas_diarias  
count     7043.000000  
mean         2.158675  
std          1.003088  
min          0.610000  
25%          1.180000  
50%          2.340000  
75%          2.990000  
max          3.960000

In [ ]:
cores = ['#67C9E0', '#E0677F']

In [ ]:
import plotly.graph_objects as go

contagem_churn = dados_normalizados['Churn'].value_counts()

fig = go.Figure(data=[go.Pie(
    labels=['Não', 'Sim'],
    values=contagem_churn.values,
    pull=[0.1, 0],
    marker=dict(colors=cores),
    hoverinfo='label+value+percent',
    textinfo='percent'
)])

fig.update_layout(title='Distribuição de Churn na empresa TelecomX',
                  width=800,
                  height=800)
fig.show()

#Detalhamento de variáveis - Geral

In [ ]:
from matplotlib.ticker import MultipleLocator
from matplotlib import figure

In [ ]:
contagem = dados_normalizados['meses_contrato'].value_counts().sort_index()
bins = list(range(0, contagem.index.max() + 5, 5))
labels_contrato = [f"{i+1}-{i+5}" for i in bins[:-1]]

contagem_df = contagem.reset_index()
contagem_df.columns = ['meses', 'quantidade']
contagem_df['faixa'] = pd.cut(contagem_df['meses'], bins=bins, labels=labels_contrato, right=True)
agrupado = contagem_df.groupby('faixa')['quantidade'].sum()

/tmp/ipython-input-33-3097798310.py:8: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



In [ ]:
import numpy as np

cobranca = dados_normalizados['cobranca_mensal'].value_counts().sort_index()

max_cobranca = cobranca.index.max()
bins_cobranca = np.arange(0, max_cobranca + 5, 5)
labels_cobranca = [f"{int(bins_cobranca[i]+0.1)}-{int(bins_cobranca[i+1])}" for i in range(len(bins_cobranca)-1)]


cobranca = cobranca.reset_index()
cobranca.columns = ['meses', 'quantidade']
cobranca['faixa'] = pd.cut(cobranca['meses'], bins=bins_cobranca, labels=labels_cobranca, right=True, include_lowest=True)
agrupado_cobranca = cobranca.groupby('faixa')['quantidade'].sum()
agrupado_cobranca.head()

/tmp/ipython-input-34-1078788363.py:13: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



faixa
0-5        0
5-10       0
10-15      0
15-20    656
20-25    750
Name: quantidade, dtype: int64

In [ ]:
gasto_total = dados_normalizados['gasto_total'].value_counts().sort_index()

bins_gasto_total = np.arange(18, 8100, 500)
labels_gasto_total = [f"{int(bins_gasto_total[i]+0.1)}-{int(bins_gasto_total[i+1])}" for i in range(len(bins_gasto_total)-1)]


gasto_total = gasto_total.reset_index()
gasto_total.columns = ['valor', 'qtd_clientes']
gasto_total['faixa'] = pd.cut(gasto_total['valor'], bins=bins_gasto_total, labels=labels_gasto_total, right=True, include_lowest=True)
agrupado_gasto_total = gasto_total.groupby('faixa')['qtd_clientes'].sum()
agrupado_gasto_total

/tmp/ipython-input-35-3680619856.py:10: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



faixa
18-518       2032
518-1018      887
1018-1518     767
1518-2018     506
2018-2518     354
2518-3018     296
3018-3518     289
3518-4018     266
4018-4518     249
4518-5018     261
5018-5518     203
5518-6018     237
6018-6518     193
6518-7018     170
7018-7518     133
7518-8018     115
Name: qtd_clientes, dtype: int64

In [ ]:
contas_diarias = dados_normalizados['contas_diarias'].value_counts().sort_index()

bins_contas_diarias = [0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4]
labels_contas_diarias = [f"{float(bins_contas_diarias[i]+0.1)}-{float(bins_contas_diarias[i+1])}" for i in range(len(bins_contas_diarias)-1)]

contas_diarias = contas_diarias.reset_index()
contas_diarias.columns = ['valor', 'qtd_clientes']
contas_diarias['faixa'] = pd.cut(contas_diarias['valor'], bins=bins_contas_diarias, labels=labels_contas_diarias, right=True, include_lowest=True)
agrupado_contas_diarias = contas_diarias.groupby('faixa')['qtd_clientes'].sum()
agrupado_contas_diarias

/tmp/ipython-input-36-2088075755.py:9: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



faixa
0.6-1.0    1659
1.1-1.5     387
1.6-2.0     883
2.1-2.5    1009
2.6-3.0    1386
3.1-3.5    1214
3.6-4.0     505
Name: qtd_clientes, dtype: int64

In [ ]:
import plotly.express as px
from plotly.subplots import make_subplots

fig = make_subplots(rows=2, cols=5, subplot_titles=['Gênero', 'Idosos', 'Parceiros', 'Dependentes', 'Tempo de contrato', 'Serviço Telefônico', 'Múltiplas Linhas',
                                                    'Serviço de Internet', 'Segurança Online', 'Backup Online'])

fig.add_trace(go.Bar(x=['Feminino', 'Masculino'],
                     y=dados_normalizados['genero'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=1, col=1)

fig.add_trace(go.Bar(x=['Não', 'Sim'],
                     y=dados_normalizados['idoso'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=1, col=2)

fig.add_trace(go.Bar(x=['Não', 'Sim'],
                     y=dados_normalizados['parceiro'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=1, col=3)

fig.add_trace(go.Bar(x=['Não', 'Sim'],
                     y=dados_normalizados['dependentes'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=1, col=4)

fig.add_trace(go.Bar(x=[0, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75],
                     y=agrupado.values,
                     hovertemplate=agrupado.index.astype(str)+', %{y}',
                     marker_color='#67C9E0'), row=1, col=5)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço telêfonico'],
                     y=dados_normalizados['servico_telefonico'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=2, col=1)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço telêfonico'],
                     y=dados_normalizados['multiplas_linhas'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=2, col=2)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço de internet'],
                     y=dados_normalizados['servico_internet'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=2, col=3)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço de internet'],
                     y=dados_normalizados['seguranca_online'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=2, col=4)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço de internet'],
                     y=dados_normalizados['backup_online'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=2, col=5)

fig.update_layout(title='Distribuição de clientes por categoria na empresa TelecomX - Geral',
                  xaxis = dict(tickangle=45),
                  width=1400,
                  height=700,
                  font_family='Arial',
                  title_font_color='black',
                  title_font_size=16,
                  font_size=12,
                  showlegend=False)
fig.show()

In [ ]:
fig = make_subplots(rows=2, cols=5, subplot_titles=['Proteção de Dispositivo', 'Suporte Técnico', 'TV a Cabo', 'Streaming de Filmes', 'Tipo de Contrato', 'Fatura Online',
                                                    'Método de Pagamento', 'Cobrança Mensal', 'Gasto Total', 'Contas Diárias'])

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço de Internet'],
                     y=dados_normalizados['protecao_dispositivo'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=1, col=1)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço de Internet'],
                     y=dados_normalizados['suporte_tecnico'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=1, col=2)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço de Internet'],
                     y=dados_normalizados['tv_a_cabo'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=1, col=3)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço de Internet'],
                     y=dados_normalizados['streaming_filmes'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=1, col=4)

fig.add_trace(go.Bar(x=['Mensal', 'Anual', 'Bienal'],
                     y=dados_normalizados['contrato'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=1, col=5)

fig.add_trace(go.Bar(x=['Não', 'Sim'],
                     y=dados_normalizados['fatura_online'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=2, col=1)

fig.add_trace(go.Bar(x=['Cheque', 'Cheque eletrônico', 'Cartão de crédito', 'Transferência Bancária'],
                     y=dados_normalizados['metodo_pagamento'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=2, col=2)

fig.add_trace(go.Bar(x=list(range(0, 120, 5)),
                     y=agrupado_cobranca.values,
                     marker_color='#67C9E0',
                     hovertemplate=agrupado_cobranca.index.astype(str)+', %{y}'), row=2, col=3)

fig.add_trace(go.Bar(x=list(range(0, 8200, 500)),
                     y=agrupado_gasto_total.values,
                     marker_color='#67C9E0',
                     hovertemplate=agrupado_gasto_total.index.astype(str)+', %{y}'), row=2, col=4)

fig.add_trace(go.Bar(x=[0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4],
                     y=agrupado_contas_diarias.values,
                     marker_color='#67C9E0',
                     hovertemplate=agrupado_contas_diarias.index.astype(str)+', %{y}'), row=2, col=5)

fig.update_layout(title='Distribuição de clientes por categoria na empresa TelecomX - Geral',
                  xaxis = dict(tickangle=45),
                  width=1400,
                  height=700,
                  font_family='Arial',
                  title_font_color='black',
                  title_font_size=16,
                  font_size=12,
                  showlegend=False)
fig.show()

#Detalhamento Variáveis - Churn x No Churn

In [ ]:
dados_churn = dados_normalizados.query('Churn == 1')
dados_no_churn = dados_normalizados.query('Churn == 0')

In [ ]:
#Churn

contagem_churn = dados_churn['meses_contrato'].value_counts().sort_index()
bins_churn = list(range(0, contagem_churn.index.max() + 5, 5))
labels_contrato_churn = [f"{i+1}-{i+5}" for i in bins[:-1]]
contagem_churn = contagem_churn.reset_index()
contagem_churn.columns = ['meses', 'quantidade']
contagem_churn['faixa'] = pd.cut(contagem_churn['meses'], bins=bins_churn, labels=labels_contrato_churn, right=True)
agrupado_churn = contagem_churn.groupby('faixa')['quantidade'].sum()

cobranca_churn = dados_churn['cobranca_mensal'].value_counts().sort_index()
max_cobranca_churn = cobranca_churn.index.max()
bins_cobranca_churn = np.arange(0, max_cobranca_churn + 5, 5)
labels_cobranca_churn = [f"{int(bins_cobranca_churn[i]+0.1)}-{int(bins_cobranca_churn[i+1])}" for i in range(len(bins_cobranca_churn)-1)]
cobranca_churn = cobranca_churn.reset_index()
cobranca_churn.columns = ['meses', 'quantidade']
cobranca_churn['faixa'] = pd.cut(cobranca_churn['meses'], bins=bins_cobranca_churn, labels=labels_cobranca_churn, right=True, include_lowest=True)
agrupado_cobranca_churn = cobranca_churn.groupby('faixa')['quantidade'].sum()

gasto_total_churn = dados_churn['gasto_total'].value_counts().sort_index()
bins_gasto_total_churn = np.arange(18, 8100, 500)
labels_gasto_total_churn = [f"{int(bins_gasto_total_churn[i]+0.1)}-{int(bins_gasto_total_churn[i+1])}" for i in range(len(bins_gasto_total_churn)-1)]
gasto_total_churn = gasto_total_churn.reset_index()
gasto_total_churn.columns = ['valor', 'qtd_clientes']
gasto_total_churn['faixa'] = pd.cut(gasto_total_churn['valor'], bins=bins_gasto_total_churn, labels=labels_gasto_total_churn, right=True, include_lowest=True)
agrupado_gasto_total_churn = gasto_total_churn.groupby('faixa')['qtd_clientes'].sum()

contas_diarias_churn = dados_churn['contas_diarias'].value_counts().sort_index()
bins_contas_diarias_churn = [0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4]
labels_contas_diarias_churn = [f"{float(bins_contas_diarias_churn[i]+0.1)}-{float(bins_contas_diarias_churn[i+1])}" for i in range(len(bins_contas_diarias_churn)-1)]
contas_diarias_churn = contas_diarias_churn.reset_index()
contas_diarias_churn.columns = ['valor', 'qtd_clientes']
contas_diarias_churn['faixa'] = pd.cut(contas_diarias_churn['valor'], bins=bins_contas_diarias_churn, labels=labels_contas_diarias_churn, right=True, include_lowest=True)
agrupado_contas_diarias_churn = contas_diarias_churn.groupby('faixa')['qtd_clientes'].sum()

/tmp/ipython-input-40-3804313740.py:9: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.

/tmp/ipython-input-40-3804313740.py:18: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.

/tmp/ipython-input-40-3804313740.py:26: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.

/tmp/ipython-input-40-3804313740.py:34: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass obs

In [ ]:
#No Churn

contagem_no_churn = dados_no_churn['meses_contrato'].value_counts().sort_index()
bins_no_churn = list(range(0, contagem_no_churn.index.max() + 5, 5))
labels_contrato_no_churn = [f"{i+1}-{i+5}" for i in bins[:-1]]
contagem_no_churn = contagem_no_churn.reset_index()
contagem_no_churn.columns = ['meses', 'quantidade']
contagem_no_churn['faixa'] = pd.cut(contagem_no_churn['meses'], bins=bins_no_churn, labels=labels_contrato_no_churn, right=True)
agrupado_no_churn = contagem_no_churn.groupby('faixa')['quantidade'].sum()

cobranca_no_churn = dados_no_churn['cobranca_mensal'].value_counts().sort_index()
max_cobranca_no_churn = cobranca_no_churn.index.max()
bins_cobranca_no_churn = np.arange(0, max_cobranca_no_churn + 5, 5)
labels_cobranca_no_churn = [f"{int(bins_cobranca_no_churn[i]+0.1)}-{int(bins_cobranca_no_churn[i+1])}" for i in range(len(bins_cobranca_no_churn)-1)]
cobranca_no_churn = cobranca_no_churn.reset_index()
cobranca_no_churn.columns = ['meses', 'quantidade']
cobranca_no_churn['faixa'] = pd.cut(cobranca_no_churn['meses'], bins=bins_cobranca_no_churn, labels=labels_cobranca_no_churn, right=True, include_lowest=True)
agrupado_cobranca_no_churn = cobranca_no_churn.groupby('faixa')['quantidade'].sum()

gasto_total_no_churn = dados_no_churn['gasto_total'].value_counts().sort_index()
bins_gasto_total_no_churn = np.arange(18, 8100, 500)
labels_gasto_total_no_churn = [f"{int(bins_gasto_total_no_churn[i]+0.1)}-{int(bins_gasto_total_no_churn[i+1])}" for i in range(len(bins_gasto_total_no_churn)-1)]
gasto_total_no_churn = gasto_total_no_churn.reset_index()
gasto_total_no_churn.columns = ['valor', 'qtd_clientes']
gasto_total_no_churn['faixa'] = pd.cut(gasto_total_no_churn['valor'], bins=bins_gasto_total_no_churn, labels=labels_gasto_total_no_churn, right=True, include_lowest=True)
agrupado_gasto_total_no_churn = gasto_total_no_churn.groupby('faixa')['qtd_clientes'].sum()

contas_diarias_no_churn = dados_no_churn['contas_diarias'].value_counts().sort_index()
bins_contas_diarias_no_churn = [0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4]
labels_contas_diarias_no_churn = [f"{float(bins_contas_diarias_no_churn[i]+0.1)}-{float(bins_contas_diarias_no_churn[i+1])}" for i in range(len(bins_contas_diarias_no_churn)-1)]
contas_diarias_no_churn = contas_diarias_no_churn.reset_index()
contas_diarias_no_churn.columns = ['valor', 'qtd_clientes']
contas_diarias_no_churn['faixa'] = pd.cut(contas_diarias_no_churn['valor'], bins=bins_contas_diarias_no_churn, labels=labels_contas_diarias_no_churn, right=True, include_lowest=True)
agrupado_contas_diarias_no_churn = contas_diarias_no_churn.groupby('faixa')['qtd_clientes'].sum()

/tmp/ipython-input-41-1800933316.py:9: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.

/tmp/ipython-input-41-1800933316.py:18: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.

/tmp/ipython-input-41-1800933316.py:26: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.

/tmp/ipython-input-41-1800933316.py:34: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass obs

In [ ]:
fig = make_subplots(rows=2, cols=5, subplot_titles=['Gênero', 'Idosos', 'Parceiros', 'Dependentes', 'Tempo de contrato', 'Serviço Telefônico', 'Múltiplas Linhas',
                                                    'Serviço de Internet', 'Segurança Online', 'Backup Online'])

fig.add_trace(go.Bar(x=['Feminino', 'Masculino'],
                     y=dados_churn['genero'].value_counts(),
                     marker_color='#E0677F',
                     width=0.4), row=1, col=1)

fig.add_trace(go.Bar(x=['Não', 'Sim'],
                     y=dados_churn['idoso'].value_counts(),
                     marker_color='#E0677F',
                     width=0.4), row=1, col=2)

fig.add_trace(go.Bar(x=['Não', 'Sim'],
                     y=dados_churn['parceiro'].value_counts(),
                     marker_color='#E0677F',
                     width=0.4), row=1, col=3)

fig.add_trace(go.Bar(x=['Não', 'Sim'],
                     y=dados_churn['dependentes'].value_counts(),
                     marker_color='#E0677F',
                     width=0.4), row=1, col=4)

fig.add_trace(go.Bar(x=[0, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75],
                     y=agrupado_churn.values,
                     hovertemplate=agrupado.index.astype(str)+', %{y}',
                     marker_color='#E0677F'), row=1, col=5)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço telêfonico'],
                     y=dados_churn['servico_telefonico'].value_counts(),
                     marker_color='#E0677F',
                     width=0.4), row=2, col=1)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço telêfonico'],
                     y=dados_churn['multiplas_linhas'].value_counts(),
                     marker_color='#E0677F',
                     width=0.4), row=2, col=2)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço de internet'],
                     y=dados_churn['servico_internet'].value_counts(),
                     marker_color='#E0677F',
                     width=0.4), row=2, col=3)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço de internet'],
                     y=dados_churn['seguranca_online'].value_counts(),
                     marker_color='#E0677F',
                     width=0.4), row=2, col=4)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço de internet'],
                     y=dados_churn['backup_online'].value_counts(),
                     marker_color='#E0677F',
                     width=0.4), row=2, col=5)

fig.add_trace(go.Bar(x=['Feminino', 'Masculino'],
                     y=dados_no_churn['genero'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=1, col=1)

fig.add_trace(go.Bar(x=['Não', 'Sim'],
                     y=dados_no_churn['idoso'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=1, col=2)

fig.add_trace(go.Bar(x=['Não', 'Sim'],
                     y=dados_no_churn['parceiro'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=1, col=3)

fig.add_trace(go.Bar(x=['Não', 'Sim'],
                     y=dados_no_churn['dependentes'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=1, col=4)

fig.add_trace(go.Bar(x=[0, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75],
                     y=agrupado_no_churn.values,
                     hovertemplate=agrupado.index.astype(str)+', %{y}',
                     marker_color='#67C9E0'), row=1, col=5)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço telêfonico'],
                     y=dados_no_churn['servico_telefonico'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=2, col=1)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço telêfonico'],
                     y=dados_no_churn['multiplas_linhas'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=2, col=2)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço de internet'],
                     y=dados_no_churn['servico_internet'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=2, col=3)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço de internet'],
                     y=dados_no_churn['seguranca_online'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=2, col=4)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço de internet'],
                     y=dados_no_churn['backup_online'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=2, col=5)

fig.update_layout(title='Distribuição de clientes por categoria na empresa TelecomX - Churn x No Churn',
                  xaxis = dict(tickangle=45),
                  barmode='group',
                  width=1400,
                  height=700,
                  font_family='Arial',
                  title_font_color='black',
                  title_font_size=16,
                  font_size=12,
                  showlegend=False)
fig.show()

In [ ]:
fig = make_subplots(rows=2, cols=5, subplot_titles=['Proteção de Dispositivo', 'Suporte Técnico', 'TV a Cabo', 'Streaming de Filmes', 'Tipo de Contrato', 'Fatura Online',
                                                    'Método de Pagamento', 'Cobrança Mensal', 'Gasto Total', 'Contas Diárias'])

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço de Internet'],
                     y=dados_churn['protecao_dispositivo'].value_counts(),
                     marker_color='#E0677F',
                     width=0.4), row=1, col=1)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço de Internet'],
                     y=dados_churn['suporte_tecnico'].value_counts(),
                     marker_color='#E0677F',
                     width=0.4), row=1, col=2)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço de Internet'],
                     y=dados_churn['tv_a_cabo'].value_counts(),
                     marker_color='#E0677F',
                     width=0.4), row=1, col=3)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço de Internet'],
                     y=dados_churn['streaming_filmes'].value_counts(),
                     marker_color='#E0677F',
                     width=0.4), row=1, col=4)

fig.add_trace(go.Bar(x=['Mensal', 'Anual', 'Bienal'],
                     y=dados_churn['contrato'].value_counts(),
                     marker_color='#E0677F',
                     width=0.4), row=1, col=5)

fig.add_trace(go.Bar(x=['Não', 'Sim'],
                     y=dados_churn['fatura_online'].value_counts(),
                     marker_color='#E0677F',
                     width=0.4), row=2, col=1)

fig.add_trace(go.Bar(x=['Cheque', 'Cheque eletrônico', 'Cartão de crédito', 'Transferência Bancária'],
                     y=dados_churn['metodo_pagamento'].value_counts(),
                     marker_color='#E0677F',
                     width=0.4), row=2, col=2)

fig.add_trace(go.Bar(x=list(range(0, 120, 5)),
                     y=agrupado_cobranca_churn.values,
                     marker_color='#E0677F',
                     hovertemplate=agrupado_cobranca_churn.index.astype(str)+', %{y}'), row=2, col=3)

fig.add_trace(go.Bar(x=list(range(0, 8200, 500)),
                     y=agrupado_gasto_total_churn.values,
                     marker_color='#E0677F',
                     hovertemplate=agrupado_gasto_total_churn.index.astype(str)+', %{y}'), row=2, col=4)

fig.add_trace(go.Bar(x=[0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4],
                     y=agrupado_contas_diarias_churn.values,
                     marker_color='#E0677F',
                     hovertemplate=agrupado_contas_diarias_churn.index.astype(str)+', %{y}'), row=2, col=5)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço de Internet'],
                     y=dados_no_churn['protecao_dispositivo'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=1, col=1)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço de Internet'],
                     y=dados_no_churn['suporte_tecnico'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=1, col=2)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço de Internet'],
                     y=dados_no_churn['tv_a_cabo'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=1, col=3)

fig.add_trace(go.Bar(x=['Não', 'Sim', 'Sem serviço de Internet'],
                     y=dados_no_churn['streaming_filmes'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=1, col=4)

fig.add_trace(go.Bar(x=['Mensal', 'Anual', 'Bienal'],
                     y=dados_no_churn['contrato'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=1, col=5)

fig.add_trace(go.Bar(x=['Não', 'Sim'],
                     y=dados_no_churn['fatura_online'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=2, col=1)

fig.add_trace(go.Bar(x=['Cheque', 'Cheque eletrônico', 'Cartão de crédito', 'Transferência Bancária'],
                     y=dados_no_churn['metodo_pagamento'].value_counts(),
                     marker_color='#67C9E0',
                     width=0.4), row=2, col=2)

fig.add_trace(go.Bar(x=list(range(0, 120, 5)),
                     y=agrupado_cobranca_no_churn.values,
                     marker_color='#67C9E0',
                     hovertemplate=agrupado_cobranca_no_churn.index.astype(str)+', %{y}'), row=2, col=3)

fig.add_trace(go.Bar(x=list(range(0, 8200, 500)),
                     y=agrupado_gasto_total_no_churn.values,
                     marker_color='#67C9E0',
                     hovertemplate=agrupado_gasto_total_no_churn.index.astype(str)+', %{y}'), row=2, col=4)

fig.add_trace(go.Bar(x=[0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4],
                     y=agrupado_contas_diarias_no_churn.values,
                     marker_color='#67C9E0',
                     hovertemplate=agrupado_contas_diarias_no_churn.index.astype(str)+', %{y}'), row=2, col=5)

fig.update_layout(title='Distribuição de clientes por categoria na empresa TelecomX - Churn x No Churn',
                  xaxis = dict(tickangle=45),
                  barmode='group',
                  width=1400,
                  height=700,
                  font_family='Arial',
                  title_font_color='black',
                  title_font_size=16,
                  font_size=12,
                  showlegend=False)
fig.show()

#📄Relatorio Final

# **Introdução**

O desafio 'TelecomX' tem como objetivo analisar os dados de uma empresa de telecomunicações para identificar os principais fatores que contribuem para a evasão de clientes. No desenvolvimento desse projeto, realizei o processo de ETL (Extract, Transform and Load) dos dados, seguido da criação de visualizações analíticas que possibilitam compreender o perfil dos clientes que mais cancelam os serviços. A partir dos insights obtidos, elaborei recomendações estratégicas que podem ajudar a empresa a reter seus atuais clientes e atrair novos de forma mais eficaz.

# 🧹**Limpeza e Tratamento dos Dados**
*   Leitura, normalização e limpeza dos dados: Importei bibliotecas e fiz a leitura do arquivo através da url. Como o arquivo está em JSON houve a necessidade de normalizar o arquivo utilizando a biblioteca request que transforma esses dados em um dataframe, facilitando a análise.

*   Verificando inconsistências: função info() para verificar os tipos de dados do dataframe, isnull() para verificar dados nulos, unique() para verificar os valores atribuídos ao dataframe, duplicate() para verificar dados duplicados, função lambda para entender se existem valores vazios ('') no dataframe.

*   Tratando as inconsistências: alteração dos nomes das colunas para onde houvesse '.' fosse substituído por '_'. Analisei os dados onde a coluna Churn não apresenta valor para verificar se existem padrões nos dados ausentes, verificando que não existe padrão e entendendo que esses dados representam 3% de todo o dataframe optei por excluir essas linhas para conseguir fazer análises mais assertivas sem ser prejudicada por esses dados ausentes. Alterei a coluna 'account.Charges.Total' para float para tratar os valores vazios que possuíam nessa colunas.

*   Outros: remoção da coluna 'customerID'. Reset no index do dataframe, alteração de 'yes' para 1 e 'no' para 0 para melhorar o processamento matemático e criar uma padronização no dataframe. Criação da coluna contas diárias. Renomeei e traduzi os nomes das colunas e os dados no dataframe a fim de tornar os dados mais acessíveis e compreensíveis. Também deixei todas as infos em minúsculas a fim de padronização.







# **📈 Análise Exploratória dos Dados**


Análise da distribuição de evasão através do gráfico de pizza. Entendemos que a evasão de cliente qualifica cerca de 27% dos clientes, um número bem expressivo e preocupante para a empresa.

Após isso decidi verificar, em gráficos simples de barras através de subplots, todas as variáveis de forma geral. Insigths que conseguimos extrair analisando esses gráficos:

* quase 50% dos clientes possuem um parceiro,
* existem um número muito expressivo de clientes com pouco tempo de contrato(entre 1 e 5 meses),
* número muito baixo de clientes que usam serviço telefônico,
* a maior parte dos clientes possuem contrato mensal,
* a forma de pagamento mais comum é através do cheque.

Da mesma forma que feito anteriormente, uma análise através de gráficos de barras e subplots, porém agora analisando dados de Churn e No Churn lado a lado(vermelho=Churn, azul=No Churn). Insigths que conseguimos extrair analisando dessa forma:
* clientes que possuem dependentes cancelam menos,
* índice altíssimo de cancelamento em clientes com poucos meses de contrato (1 a 5 meses),
* clientes que não possuem serviços adicionais tem mais incidência em cancelamento,
* maior parte dos clientes que não possui suporte técnico adicional cancelam,
* altíssimo índice de cancelamentos em contratos mensais,
* altíssimo índice de cancelamento de clientes que pagam com cheque físico,
* analisando os gastos totais, clientes que mais cancelam são os que menos gastam.

# ✍️Conclusões e Insights

Analisando e ligando os fatos podemos concluir que:

* Clientes com contratos mensais e recentes (1 a 5 meses) possuem altíssimo índice de evasão, podendo ser indicativo de problemas na fase inicial de experiência com a empresa e também a flexibilidade do contrato mensal facilita a evasão.
* Clientes sem serviços adicionais tem maior tendência ao cancelamento (maior índice no suporte técnico).
* Pagamento por cheque físico também está associado a evasão, enquanto que os meios digitais não têm relação tão forte.
* Clientes sem dependentes ou parceiros tendem a cancelar mais, indicando que clientes que tenham com quem compartilhar seus benefícios tenham maior chances de permanecer seu contrato com a empresa.
* Clientes que menos gastam são o que mais cancelam, podendo ter relação com o tempo permanecido na empresa e também com o fato de não obter serviços adicionais.

# 💡 Recomendações

Após analisar os dados da empresa TelecomX e gerar visualizações que ajudaram a entender melhor o perfil dos clientes, seguem as principais recomendações para que a empresa consiga tratar seu problema com cancelamento e evasão de clientes.

* Procurando atingir clientes com parceiros e dependentes, ofererer planos familiares, que incluam serviços que atendem a toda a família.
* Incentivar a alteração dos contratos de atuais clientes para contratos de longo prazo e ofertar apenas contrato anuais ou bienais para novos clientes, se possível oferecendo benefícios para quem aderir esses tipos de contrato.
* Façam pacotes promocionais para adesão de serviços adicionais, se possível focar principalmente no serviço de suporte, pois estes fazem com que fidelize o cliente
* Incentivar atuais clientes a alterar sua forma de pagamento de cheque físico para algum modo de pagamento digital e ofertar apenas pagamentos digitais para novos clintes.
* Aumentar os tipos de pagamentos digitais, como PIX e boleto bancário, por exemplo.
* Aprimorar os primeiros meses de experiência do cliente com onboarding eficaz e suporte proativo nos primeiros meses de contrato, buscando assim maior fidelização.
